Add parametrized tests for `should_strip_ansi`

Tyl13 opened the pull request:
Issue #2606 described a inconsistency in how color=True works between Operating Systems. 
This is because ANSI styles can't always be applied to different terminals on windows. This did not cause issues when running it within the terminals because Colorama would change the ANSI style codes to escape codes for windows when it was required. However, if you were to pipe the `click.echo()` output to a file then the `should_strip_ansi` function would not work on Windows as the other operating systems. 

This is because the `auto_wrap_for_ansi` function was not being called with the color argument. After the `auto_wrap_for_ansi` function is called, it defaults `color=None` and will end up calling `should_strip_ansi` with `color=None` which is different from the original call to `should_strip_ansi` that was called before the wrapper function. 

By passing color to `auto_wrap_for_ansi`, then the behavior of `should_strip_ansi` becomes consistent.

An additional change applied within the PR is to the tests in `test_utils.py`. This is to increase test coverage for the different cases of `click.echo()` and removes the old version of the test. Instead of one test that mocks `isatty()`, there are now three different tests. One that mocks `isatty()=True` and `_is_jupyter_kernel_output()=False`, another that mocks `isatty()=False` and `_is_jupyter_kernel_output()=True` , and the last one mocks both `isatty()=False` and `_is_jupyter_kernel_output()=False`.

fixes #2606 

- [x] Add tests that demonstrate the correct behavior of the change. Tests
  should fail without the change.
- [x] Add or update relevant docs, in the docs folder and in code.
- [x] Add an entry in CHANGES.rst summarizing the change and linking to the issue.

Tyl13 commented:
Rebased to fix merge conflicts more easily. However, something seems to sometimes cause the windows version to fail. Currently investigating that. I've had it fail locally and then a second rerun succeed. I'm unsure what is causing the issue atm.

Tyl13 commented:
The issue is coming from my previous change to the runner. The issue is documented here #2732 with the PR to fix it here #2733.

AndreasBackx commented:
Fixed by #2607

davidism commented:
After rebasing this on latest stable that includes #2607 and #2733, the `test_echo_color_flag` that you change fails now. I have a feeling the test doesn't need to be changed if the code was changed. I'm not going to block 8.1.8 on this since it's only a change to tests. @Tyl13 if you still remember any discussion about this from the sprint, let me know, and if you can update the test to pass again that would be great.

Tyl13 commented:
Yeah, I am also not sure why I changed the `isatty` to `should_strip_ansi`. That one doesn't need changed, maybe I was trying to make it closer to the new test in the `test_compat` but I don't have my previous laptop to check my reasoning anymore. Looking at the previous commit before I force-pushed the fix-2606 branch from [56d016e](https://github.com/pallets/click/commit/56d016ef912ad860fa38756003fa62efa0c709b5) to [8e34c62](https://github.com/pallets/click/commit/8e34c623a10ad2d26137971ec340eb84a04f6212) shows the tests being very different. I then moved those to test_compat on their own, and I'm guessing renamed the `isatty` to `should_strip_ansi` with a find and replace. I'm not sure how I was getting it to pass then though, unless I noticed it on my end and forgot to commit undoing it.

I can fix it by just removing those changes to that file,do you want it rebased on the most recent tag of 8.1.8?

kdeldycke commented:
The fix consisting in passing `color` to `auto_wrap_for_ansi` shipped in 8.2.0 via #2607. So this PR is only about adding color-related tests, which are worth reusing and maybe extend.

I will try to clean-up that PR.

kdeldycke committed:
Add parametrized tests for `should_strip_ansi`

Copy of the original code from #2731, but without the part that has been addressed upstream by #2607 for 8.2.0 release.

Co-authored-by: Tyler Lendon <trl75@pitt.edu>
Co-authored-by: David Lord <davidism@gmail.com>

AndreasBackx commented:
I cannot review my own PR, but the rebases look great. Thanks again!

kdeldycke commented:
> I cannot review my own PR, but the rebases look great. Thanks again!

Just added you as a review on this one if this help.

kdeldycke commented:
This PR has been cleanup of duplicate code and rebased on top of stable and ready to be merged upstream.
